### PR TITLE
Ivanov Nikita. Lab 2. Var 3.

### PR DESCRIPTION
**Pull request by Atikin-NT**
Реализован плагин и тесты по задаче 

> _Найти все циклы внутри функции и обернуть их в вызовы loop_start()/loop_end(). Можно использовать Loop Analysis._.

### CODE REVIEW

File: llvm/test/lab2/ivanov_nikita/test.ll [hunk 1]
```diff
@@ -0,0 +1,44 @@
+; RUN: opt -passes=ivanov-loop-pass -S %s 2>&1 | FileCheck %s
+
+; void foo(int n, int m) {
+;     int c0;
+;     int c1;
+;     for (c0 = n; c0 > 0; c0--) {
```
Comment (m-ly4) [inline llvm/test/lab2/ivanov_nikita/test.ll]: Could you please add the following test cases:
- while, do-while loops
- no loops but if-else condition present (check that no insertion is done)
- function with any loop and already added `loop_*` calls

File: llvm/labs/lab2/ivanov_nikita/ivanov.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    auto &LI = FAM.getResult<LoopAnalysis>(F);
+
+    for (auto &L : LI) { // for by loop
+      auto *Preheader =
+          L->getLoopPreheader();           // get preheader block of the loop
+      auto *ExitBlock = L->getExitBlock(); // end exit block of the loop
+      IRBuilder<> Builder(Preheader->getTerminator()); // api for basic block
+
+      if (!Preheader)
```
Comment (m-ly4) [inline llvm/labs/lab2/ivanov_nikita/ivanov.cpp]: I'd suggest firstly perform all checks and if all is fine then insert `loop_*` calls.

File: llvm/labs/lab2/ivanov_nikita/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (m-ly4) [inline llvm/labs/lab2/ivanov_nikita/CMakeLists.txt]:
```suggestion
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab2/ivanov_nikita/test.ll [hunk 4]
```diff
@@ -0,0 +1,174 @@
+; RUN: opt -passes=ivanov-loop-pass -S %s 2>&1 | FileCheck %s
```
Comment (m-ly4) [inline llvm/test/lab2/ivanov_nikita/test.ll]: Could you please load plugin as dynamic library at runtime?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by Atikin-NT** [comment]
@m-ly4 I've done, you can check

File: llvm/test/lab2/ivanov_nikita/test.ll [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+; void with_lopps_func(){
+;     int i = 0;
+;     loop_start();
+;     while(i < 10){
+;         i++;
+;     } 
+;     loop_end();
+; }
+
```
Comment (m-ly4) [inline llvm/test/lab2/ivanov_nikita/test.ll:45]:
```suggestion
void LoopWithRet(int &a) {
  int i = 0;
  while(i < 100){
         i++;
         if (i > 10) {
            return;
         }
         a = a + i;
     }
     a *= i -1;
     return;
}

```

File: llvm/test/lab2/ivanov_nikita/test.ll [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+; CHECK-NOT:    call void @_Z8loop_endv()
+  ret void
+}
+
+define dso_local void @_Z15with_lopps_funcv() {
+entry:
+  %i = alloca i32, align 4
+  store i32 0, ptr %i, align 4
+; CHECK:    call void @_Z10loop_startv()
+  br label %while.cond
```
Comment (m-ly4) [inline llvm/test/lab2/ivanov_nikita/test.ll:168]:
```suggestion
; CHECK-NOT: loop_start
  br label %while.cond
```

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**